
# SUPPLY CHAIN ANALYSIS IN ASIA

## IMPORTING LIBRARIES 

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## IDENTIFICATION OF PATH

In [2]:
path = r'C:\Users\gryzu\Supply Chain Management'

In [3]:
path

'C:\\Users\\gryzu\\Supply Chain Management'

## IMPORTING DATASET

In [4]:
df_asia = pd.read_excel(os.path.join(path, '1 DATA', 'ASIASCM.xlsx'))

## DATAFRAME FIRST ANALYSIS

In [5]:
# Representing all columns 
df_asia.head()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         2                              1   
4  TRANSFER                         2                              1   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          95.180000          288.420013     Late delivery   
4          68.430000          285.140015     Late delivery   

   Late_delivery_risk  Category Id   Category Name Customer City  ...  \
0                   0           73  Sporting Goods        Caguas  ...   
1                   1           73  Sporting Goods        Caguas  ...   
2                   0           73  Sporting Goods      San Jose  ...   
3                   1           73  Sporting Goods        Caguas  ...   
4                   1           73  Sporting Goods         Miami  ...   

  Order Zipcode Product Card Id Product Category Id  Product Description  \
0           NaN            1360                  73                  NaN   
1           NaN            1360                  73                  NaN   
2           NaN            1360                  73                  NaN   
3           NaN            1360                  73                  NaN   
4           NaN            1360                  73                  NaN   

                                  Product Image  Product Name Product Price  \
0  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
1  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
2  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
3  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
4  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   

  Product Status shipping date (DateOrders)   Shipping Mode  
0              0        2018-03-02 22:56:00  Standard Class  
1              0            1/18/2018 12:27  Standard Class  
2              0            1/17/2018 12:06  Standard Class  
3              0            1/15/2018 10:42     First Class  
4              0            1/15/2018 10:21     First Class  

[5 rows x 53 columns]

In [6]:
df_asia.tail()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
31107    DEBIT                         6                              4   
31108  PAYMENT                         3                              4   
31109     CASH                         4                              4   
31110    DEBIT                         3                              2   
31111  PAYMENT                         4                              4   

       Benefit per order  Sales per customer   Delivery Status  \
31107          86.400002          319.980011     Late delivery   
31108         119.989998          299.989990  Advance shipping   
31109          40.000000          399.980011  Shipping on time   
31110        -613.770019          395.980011     Late delivery   
31111         168.949997          383.980011  Shipping on time   

       Late_delivery_risk  Category Id Category Name Customer City  ...  \
31107                   1           45       Fishing        Caguas  ...   
31108                   0           45       Fishing     Lancaster  ...   
31109                   0           45       Fishing      Brooklyn  ...   
31110                   1           45       Fishing   Bakersfield  ...   
31111                   0           45       Fishing        Caguas  ...   

      Order Zipcode Product Card Id Product Category Id  Product Description  \
31107           NaN            1004                  45                  NaN   
31108           NaN            1004                  45                  NaN   
31109           NaN            1004                  45                  NaN   
31110           NaN            1004                  45                  NaN   
31111           NaN            1004                  45                  NaN   

                                           Product Image  \
31107  http://images.acmesports.sports/Field+%26+Stre...   
31108  http://images.acmesports.sports/Field+%26+Stre...   
31109  http://images.acmesports.sports/Field+%26+Stre...   
31110  http://images.acmesports.sports/Field+%26+Stre...   
31111  http://images.acmesports.sports/Field+%26+Stre...   

                                    Product Name Product Price Product Status  \
31107  Field & Stream Sportsman 16 Gun Fire Safe    399.980011              0   
31108  Field & Stream Sportsman 16 Gun Fire Safe    399.980011              0   
31109  Field & Stream Sportsman 16 Gun Fire Safe    399.980011              0   
31110  Field & Stream Sportsman 16 Gun Fire Safe    399.980011              0   
31111  Field & Stream Sportsman 16 Gun Fire Safe    399.980011              0   

      shipping date (DateOrders)   Shipping Mode  
31107             1/22/2016 6:07  Standard Class  
31108             1/19/2016 5:04  Standard Class  
31109             1/20/2016 3:40  Standard Class  
31110             1/19/2016 1:34    Second Class  
31111            1/19/2016 18:54  Standard Class  

[5 rows x 53 columns]

In [7]:
# Representing all variables 
df_asia.columns

Index(['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Id',
       'Customer Lname', 'Customer Password', 'Customer Segment',
       'Customer State', 'Customer Street', 'Customer Zipcode',
       'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market',
       'Order City', 'Order Country', 'Order Customer Id',
       'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id',
       'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Order Zipcode', 'Product Card Id', 'Product Category Id',
       'Product De

We cam see some columns, which has sensitive personal data. It is necessary to drop them. Also, we can see some variables,
that are also unnecessary for our analysis (NaNs or Product Image Link)

## DATA CLEANING

#### DROPPING UNNECESSARY COLUMNS


In [8]:
# List of columns to drop
columns_to_drop = [
    'Customer City', 'Customer Country', 'Customer Email', 
    'Customer Fname', 'Customer Id', 'Customer Lname', 
    'Customer Password', 'Customer Segment', 'Customer State', 
    'Customer Street', 'Customer Zipcode', 'Latitude', 
    'Longitude', 'Order Id', 'Order Item Id', 
    'Order Customer Id', 'Order Item Cardprod Id', 
    'Order Zipcode', 'Product Image', 'Product Status'
]


In [9]:
# Drop the specified columns
df_asia_cleaned = df_asia.drop(columns=columns_to_drop)

Since the error occured in Jupyther everytime, I decided to make dropping in Excel and reimport the Dataset Again.

#### IMPORTED ACTUALIZED DATASET WITH DROPPED COLUMNS

In [10]:
df_asia_drop = pd.read_excel(os.path.join(path, '1 DATA', 'ASIASCM_drop.xlsx'))

In [11]:
# Representing an updated dataset
df_asia_drop.head()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         2                              1   
4  TRANSFER                         2                              1   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          95.180000          288.420013     Late delivery   
4          68.430000          285.140015     Late delivery   

   Late_delivery_risk  Category Id   Category Name  Department Id  ...  \
0                   0           73  Sporting Goods              2  ...   
1                   1           73  Sporting Goods              2  ...   
2                   0           73  Sporting Goods              2  ...   
3                   1           73  Sporting Goods              2  ...   
4                   1           73  Sporting Goods              2  ...   

  Order Item Total Order Profit Per Order    Order Region      Order State  \
0       314.640015              91.250000  Southeast Asia  Java Occidental   
1       311.359985            -249.089996      South Asia         Rajastán   
2       309.720001            -247.779999      South Asia         Rajastán   
3       288.420013              95.180000    Eastern Asia        Guangdong   
4       285.140015              68.430000    Eastern Asia        Guangdong   

  Order Status  Product Category Id  Product Name  Product Price  \
0     COMPLETE                   73  Smart watch          327.75   
1      PENDING                   73  Smart watch          327.75   
2       CLOSED                   73  Smart watch          327.75   
3     COMPLETE                   73  Smart watch          327.75   
4   PROCESSING                   73  Smart watch          327.75   

   shipping date (DateOrders)   Shipping Mode  
0         2018-03-02 22:56:00  Standard Class  
1             1/18/2018 12:27  Standard Class  
2             1/17/2018 12:06  Standard Class  
3             1/15/2018 10:42     First Class  
4             1/15/2018 10:21     First Class  

[5 rows x 29 columns]

In [12]:
# Representing all variables 
df_asia_drop.columns

Index(['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Department Id',
       'Department Name', 'Market', 'Order City', 'Order Country',
       'order date (DateOrders)', 'Order Item Product Price',
       'Order Item Profit Ratio', 'Order Item Quantity', 'Sales',
       'Order Item Total', 'Order Profit Per Order', 'Order Region',
       'Order State', 'Order Status', 'Product Category Id', 'Product Name',
       'Product Price', 'shipping date (DateOrders)', 'Shipping Mode'],
      dtype='object')

In [13]:
# Renaming a column 
df_asia_drop.rename(columns = {'shipping date (DateOrders)' : 'Shipping Date'}, inplace = True)

In [14]:
df_asia_drop.rename(columns = {'order date (DateOrders)' : 'Order Date'}, inplace = True)

In [15]:
df_asia_drop.rename(columns = {'Late_delivery_risk' : 'Late Delivery Risk'}, inplace = True)

#### CHECKING MISSING VALUES

In [16]:
df_asia_drop.isnull().sum()

Type                             0
Days for shipping (real)         0
Days for shipment (scheduled)    0
Benefit per order                0
Sales per customer               0
Delivery Status                  0
Late Delivery Risk               0
Category Id                      0
Category Name                    0
Department Id                    0
Department Name                  0
Market                           0
Order City                       0
Order Country                    0
Order Date                       0
Order Item Product Price         0
Order Item Profit Ratio          0
Order Item Quantity              0
Sales                            0
Order Item Total                 0
Order Profit Per Order           0
Order Region                     0
Order State                      0
Order Status                     0
Product Category Id              0
Product Name                     0
Product Price                    0
Shipping Date                    0
Shipping Mode       

No missing values are found

#### CHECKING THE DUPLICATES 

In [17]:
df_dups = df_asia_drop[df_asia_drop.duplicated()]

In [18]:
df_dups

Empty DataFrame
Columns: [Type, Days for shipping (real), Days for shipment (scheduled), Benefit per order, Sales per customer, Delivery Status, Late Delivery Risk, Category Id, Category Name, Department Id, Department Name, Market, Order City, Order Country, Order Date, Order Item Product Price, Order Item Profit Ratio, Order Item Quantity, Sales, Order Item Total, Order Profit Per Order, Order Region, Order State, Order Status, Product Category Id, Product Name, Product Price, Shipping Date, Shipping Mode]
Index: []

[0 rows x 29 columns]

No duplicates have been found

#### CHECKING DATA TYPES 

In [19]:
df_asia_drop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31112 entries, 0 to 31111
Data columns (total 29 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Type                           31112 non-null  object 
 1   Days for shipping (real)       31112 non-null  int64  
 2   Days for shipment (scheduled)  31112 non-null  int64  
 3   Benefit per order              31112 non-null  float64
 4   Sales per customer             31112 non-null  float64
 5   Delivery Status                31112 non-null  object 
 6   Late Delivery Risk             31112 non-null  int64  
 7   Category Id                    31112 non-null  int64  
 8   Category Name                  31112 non-null  object 
 9   Department Id                  31112 non-null  int64  
 10  Department Name                31112 non-null  object 
 11  Market                         31112 non-null  object 
 12  Order City                     31112 non-null 

In [20]:
# changing Order Date and Shipping Date from object to timeline 
df_asia_drop['Order Date']=pd.to_datetime(df_asia_drop['Order Date'])

In [21]:
df_asia_drop['Shipping Date']=pd.to_datetime(df_asia_drop['Shipping Date'])

In [22]:
df_asia_drop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31112 entries, 0 to 31111
Data columns (total 29 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   Type                           31112 non-null  object        
 1   Days for shipping (real)       31112 non-null  int64         
 2   Days for shipment (scheduled)  31112 non-null  int64         
 3   Benefit per order              31112 non-null  float64       
 4   Sales per customer             31112 non-null  float64       
 5   Delivery Status                31112 non-null  object        
 6   Late Delivery Risk             31112 non-null  int64         
 7   Category Id                    31112 non-null  int64         
 8   Category Name                  31112 non-null  object        
 9   Department Id                  31112 non-null  int64         
 10  Department Name                31112 non-null  object        
 11  Market         

In [23]:
df_asia_drop.dtypes

Type                                     object
Days for shipping (real)                  int64
Days for shipment (scheduled)             int64
Benefit per order                       float64
Sales per customer                      float64
Delivery Status                          object
Late Delivery Risk                        int64
Category Id                               int64
Category Name                            object
Department Id                             int64
Department Name                          object
Market                                   object
Order City                               object
Order Country                            object
Order Date                       datetime64[ns]
Order Item Product Price                float64
Order Item Profit Ratio                 float64
Order Item Quantity                       int64
Sales                                   float64
Order Item Total                        float64
Order Profit Per Order                  

In [24]:
df_asia_drop.head(5)

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         2                              1   
4  TRANSFER                         2                              1   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          95.180000          288.420013     Late delivery   
4          68.430000          285.140015     Late delivery   

   Late Delivery Risk  Category Id   Category Name  Department Id  ...  \
0                   0           73  Sporting Goods              2  ...   
1                   1           73  Sporting Goods              2  ...   
2                   0           73  Sporting Goods              2  ...   
3                   1           73  Sporting Goods              2  ...   
4                   1           73  Sporting Goods              2  ...   

  Order Item Total Order Profit Per Order    Order Region      Order State  \
0       314.640015              91.250000  Southeast Asia  Java Occidental   
1       311.359985            -249.089996      South Asia         Rajastán   
2       309.720001            -247.779999      South Asia         Rajastán   
3       288.420013              95.180000    Eastern Asia        Guangdong   
4       285.140015              68.430000    Eastern Asia        Guangdong   

  Order Status  Product Category Id  Product Name  Product Price  \
0     COMPLETE                   73  Smart watch          327.75   
1      PENDING                   73  Smart watch          327.75   
2       CLOSED                   73  Smart watch          327.75   
3     COMPLETE                   73  Smart watch          327.75   
4   PROCESSING                   73  Smart watch          327.75   

        Shipping Date   Shipping Mode  
0 2018-03-02 22:56:00  Standard Class  
1 2018-01-18 12:27:00  Standard Class  
2 2018-01-17 12:06:00  Standard Class  
3 2018-01-15 10:42:00     First Class  
4 2018-01-15 10:21:00     First Class  

[5 rows x 29 columns]

#### CHECKING OUTLIERS

In [26]:
df_asia_drop.describe()

Days for shipping (real)  Days for shipment (scheduled)  \
count              31112.000000                   31112.000000   
mean                   3.501864                       2.928227   
min                    0.000000                       0.000000   
25%                    2.000000                       2.000000   
50%                    3.000000                       4.000000   
75%                    5.000000                       4.000000   
max                    6.000000                       4.000000   
std                    1.620299                       1.369959   

       Benefit per order  Sales per customer  Late Delivery Risk  \
count       31112.000000        31112.000000        31112.000000   
mean           21.093964          180.774817            0.553806   
min         -2592.000000            7.490000            0.000000   
25%             5.420000           97.959999            0.000000   
50%            29.250000          163.770004            1.000000   
75%            63.705001          247.399994            1.000000   
max           684.000000         1500.000000            1.000000   
std           103.942982          123.989226            0.497105   

        Category Id  Department Id                     Order Date  \
count  31112.000000   31112.000000                          31112   
mean      35.616547       5.609636  2016-06-26 05:58:07.521213696   
min        3.000000       2.000000            2015-01-11 00:16:00   
25%       18.000000       4.000000            2015-12-22 17:05:45   
50%       36.000000       5.000000            2016-03-12 10:13:00   
75%       46.000000       7.000000            2016-11-13 12:42:00   
max       76.000000      12.000000            2018-12-01 22:48:00   
std       18.857814       1.861662                            NaN   

       Order Item Product Price  Order Item Profit Ratio  Order Item Quantity  \
count              31112.000000             31112.000000         31112.000000   
mean                 145.024167                 0.116595             2.032978   
min                    9.990000                -2.750000             1.000000   
25%                   49.980000                 0.070000             1.000000   
50%                   67.500000                 0.270000             1.000000   
75%                  199.990005                 0.360000             3.000000   
max                 1500.000000                 0.500000             5.000000   
std                  144.624376                 0.469347             1.423097   

              Sales  Order Item Total  Order Profit Per Order  \
count  31112.000000      31112.000000            31112.000000   
mean     201.114989        180.774817               21.093964   
min        9.990000          7.490000            -2592.000000   
25%      119.970001         97.959999                5.420000   
50%      199.919998        163.770004               29.250000   
75%      299.950012        247.399994               63.705001   
max     1500.000000       1500.000000              684.000000   
std      136.615833        123.989226              103.942982   

       Product Category Id  Product Price                  Shipping Date  
count         31112.000000   31112.000000                          31112  
mean             35.616547     145.024167  2016-06-29 17:12:22.393931776  
min               3.000000       9.990000            2015-01-11 00:06:00  
25%              18.000000      49.980000            2015-12-26 08:20:45  
50%              36.000000      67.500000            2016-03-16 13:34:00  
75%              46.000000     199.990005            2016-11-16 11:17:30  
max              76.000000    1500.000000            2018-12-01 23:51:00  
std              18.857814     144.624376                            NaN

No super heavy outliers can be found. Minimum for Benefit per Order is -2592 seems to be a little strange, however it can be due to specific type of evaluation.

## EXPORTING A CLEANED DATASET

In [25]:
# exporting the dataset as pickle
df_asia_drop.to_pickle(os.path.join(path, '1 DATA', 'FORMATTED', 'ASIASCM_CLEANED.pkl'))